client/java/src/main/java/com/linecorp/centraldogma/client/MappingWatcher.java

```diff
@@ -0,0 +1,179 @@
+            new CopyOnWriteArrayList<>();
+
+    @Nullable
+    private volatile Latest<U> mappedLatest;
+    private volatile boolean closed;
+
+    MappingWatcher(Watcher<T> parent, Function<? super T, ? extends U> mapper, Executor mapperExecutor,
+                   boolean closeParentWhenClosing) {
+        this.parent = parent;
+        this.mapper = mapper;
+        this.closeParentWhenClosing = closeParentWhenClosing;
+        parent.initialValueFuture().exceptionally(cause -> {
+            initialValueFuture.completeExceptionally(cause);
+            return null;
+        });
+        parent.watch((revision, value) -> {
+            if (closed) {
+                return;
+            }
+            final U mappedValue = mapper.apply(value);
+            if (mappedValue == null) {
+                logger.warn("mapper.apply() returned null. mapper: {}", mapper);
+                return;
+            }
+            final Latest<U> oldLatest = mappedLatest;
+            if (oldLatest != null && oldLatest.value() == mappedValue) {
+                return;
+            }
+
+            final Latest<U> newLatest = new Latest<>(revision, mappedValue);
+            mappedLatest = newLatest;
+            notifyListeners();
+            if (!initialValueFuture.isDone()) {
+                initialValueFuture.complete(newLatest);
+            }
+        }, mapperExecutor);
+    }
+
+    private void notifyListeners() {
+        if (closed) {
+            return;
+        }
+
+        final Latest<U> latest = mappedLatest;
+        assert latest != null;
+        for (Map.Entry<BiConsumer<? super Revision, ? super U>, Executor> entry : updateListeners) {
+            final BiConsumer<? super Revision, ? super U> listener = entry.getKey();
+            final Executor executor = entry.getValue();
+            executor.execute(() -> {
+                try {
+                    listener.accept(latest.revision(), latest.value());
+                } catch (Exception e) {
+                    logger.warn("Unexpected exception is raised from {}: rev={}",
+                                listener, latest.revision(), e);
+                }
+            });
+        }
+    }
+
+    @Override
+    public ScheduledExecutorService watchScheduler() {
+        return parent.watchScheduler();
+    }
+
+    @Override
+    public CompletableFuture<Latest<U>> initialValueFuture() {
+        return initialValueFuture;
+    }
+
+    @Override
+    public Latest<U> latest() {
+        final Latest<U> mappedLatest = this.mappedLatest;
+        if (mappedLatest == null) {
+            throw new IllegalStateException("value not available yet");
+        }
+        return mappedLatest;
+    }
+
+    @Override
+    public void close() {
+        closed = true;
+        if (!initialValueFuture.isDone()) {
+            initialValueFuture.cancel(false);
+        }
+        if (closeParentWhenClosing) {
```

Comment:
On second thought, I'm concerned that users may forget to close the `DefaultWatcher`.

i.e.
```
final Watcher<Boolean> watcher = dogma.client()
                                      .forRepo("foo", "bar")
                                      .watch(Query.ofText("/baz.txt")).forever().newChild(s -> s).map(s -> true);
// default watcher won't be closed
watcher.close();
```

I'm unsure if there's a way we can enforce users to avoid this pattern though. If this becomes too confusing, I think just dropping support for `newChild` may also be an option 😅 (a breaking change may be better than having to investigate potential leaks..)

Reply:
Let's continue to discuss on there: https://github.com/line/centraldogma/pull/651#discussion_r774466273